tech-specs/src/specs/ledger.md

```diff
@@ -0,0 +1,190 @@
+
+
+The WASM instantiation, the types, instructions, validation and execution of WASM modules MUST conform to the [WebAssembly specification](https://webassembly.github.io/spec/core/intro/index.html).
+
+##### WASM validation
+
+The WebAssembly code is REQUIRED to only use deterministic instructions. Furthermore, it MUST NOT use features from any of the following WebAssembly proposals:
+
+- The reference types proposal
+- The multi-value proposal
+- The bulk memory operations proposal
+- The module linking proposal
+- The SIMD proposal
+- The threads proposal
+- The tail-call proposal
+- The multi memory proposal
+- The exception handling proposal
+- The memory64 proposal
+
+##### Stack height limiter
+
+To make stack overflows deterministic, set the upper bound of the stack size to [`WASM_STACK_LIMIT`](#wasm-constants). If the stack height exceeds the limit then execution MUST abort.
+
+<!--
+cargo test test_tx_stack_limiter
+cargo test test_vp_stack_limiter
+-->
+
+##### WASM memory
+
+- TODO memory read/write gas costs
+
+##### Transaction host environment functions
+
+The following functions from the host ledger are made available in transaction's WASM code. They MAY be imported in the WASM module as shown bellow and MUST be provided by the ledger's WASM runtime:
+
+```wat
+(import "env" "gas" (func (param i32)))
+(import "env" "anoma_tx_read" (func (param i64 i64) (result i64)))
+(import "env" "anoma_tx_result_buffer" (func (param i64)))
+(import "env" "anoma_tx_has_key" (func (param i64 i64) (result i64)))
+(import "env" "anoma_tx_write" (func (param i64 i64 i64 i64)))
+(import "env" "anoma_tx_delete" (func (param i64 i64)))
+(import "env" "anoma_tx_iter_prefix" (func (param i64 i64) (result i64)))
+(import "env" "anoma_tx_iter_next" (func (param i64) (result i64)))
+(import "env" "anoma_tx_insert_verifier" (func (param i64 i64)))
+(import "env" "anoma_tx_update_validity_predicate" (func (param i64 i64 i64 i64)))
+(import "env" "anoma_tx_init_account" (func (param i64 i64 i64)))
+(import "env" "anoma_tx_get_chain_id" (func (param i64)))
+(import "env" "anoma_tx_get_block_height" (func (param ) (result i64)))
+(import "env" "anoma_tx_get_block_hash" (func (param i64)))
+(import "env" "anoma_tx_log_string" (func (param i64 i64)))
+```
+
+Additionally, the WASM module MUST export its memory as shown:
+
+```wat
+(export "memory" (memory 0))
+```
+
+- `anoma_tx_init_account` TODO newly created accounts' validity predicates aren't used until the block is committed (i.e. only the transaction that created the account may write into its storage in the block in which its being applied).
+- TODO describe functions in detail
+
+##### Validity predicate host environment functions
+
+The following functions from the host ledger are made available in validity predicate's WASM code. They MAY be imported in the WASM module as shown bellow and MUST be provided by the ledger's WASM runtime.
```

Comment:
Can we explain a bit about what each of these functions do?

Reply:
Sure, I'll describe these in detail in a follow up. (left a `TODO` placeholder)